File: content/snippets/go/interview_lcci_17_06_number_of_2s_in_range.go
package main

import "fmt"

func numberOf2sInRange(n int) int {
	count := 0
	for m := 1; m <= n; m *= 10 {
		a, b := n/m, n%m
		digit := (a % 10)
		if digit > 2 {
			count += (a/10 + 1) * m
		} else if digit == 2 {
			count += (a/10)*m + b + 1
		} else {
			count += (a / 10) * m
		}
	}
	return count
}

func main() {
	fmt.Println(numberOf2sInRange(25)) // 9
	fmt.Println(numberOf2sInRange(20)) // 2
}

File: content/snippets/typescript/interview_lcci_17_06_number_of_2s_in_range.ts
function numberOf2sInRange(n: number): number {
    let count = 0;
    for (let m = 1; m <= n; m *= 10) {
        const a = Math.trunc(n / m);
        const b = n % m;
        const digit = a % 10;
        if (digit > 2) {
            count += (Math.trunc(a / 10) + 1) * m;
        } else if (digit === 2) {
            count += Math.trunc(a / 10) * m + b + 1;
        } else {
            count += Math.trunc(a / 10) * m;
        }
    }
    return count;
}

console.log(numberOf2sInRange(25)); // 9
console.log(numberOf2sInRange(20)); // 2
